{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_RG_border_prob(mat_p, border_p, region_s, genome=None, title=\"TITLE\", merge_hom=True, line_title=None, pm_args={}, prob_args={}):\n",
    "    assert genome is not None, \"genome must be specified\"\n",
    "    region = Region(region_s, genome=genome).r\n",
    "    # --- parse border data --- #\n",
    "    if isinstance(border_p, str) or isinstance(border_p, Path):\n",
    "        data = pd.read_parquet(\n",
    "            border_p\n",
    "        )\n",
    "        data = (data == \"boundary\").astype(int)\n",
    "        data = data.sort_index()\n",
    "    else: # assume dataframe\n",
    "        data = border_p.copy()\n",
    "        data.index = pd.MultiIndex.from_product(\n",
    "            [[chrom_rm_suffix(region[0][0]) if merge_hom else region[0]], data.index]\n",
    "        )\n",
    "        #print(data)\n",
    "        data.index.names = [\"chrom\", \"start\"]\n",
    "    if merge_hom:\n",
    "        # don't separate homologous chromosomes\n",
    "        # in tad borders\n",
    "        tad_region_0 = (chrom_rm_suffix(region[0][0]),region[0][1])\n",
    "        tad_region_1 = (chrom_rm_suffix(region[1][0]),region[1][1])\n",
    "    else:\n",
    "        tad_region_0 = region[0]\n",
    "        tad_region_1 = region[1]\n",
    "    #print(tad_region_0, tad_region_1)\n",
    "    prob = data.loc[\n",
    "        tad_region_0:tad_region_1\n",
    "        :\n",
    "    ].sum(axis=1) / data.shape[1]\n",
    "    prob = prob.rename(\"TAD_prob\").reset_index()\n",
    "    if line_title is None:\n",
    "        line_title = f\"Probability of scTLD boundary\"\n",
    "    # --- generate pm --- #\n",
    "    if isinstance(mat_p, str) or isinstance(mat_p, Path):\n",
    "        pm = cool2mat(mat_p, region_s)\n",
    "        heatmap_title = \"Proximity Matrix\"\n",
    "    elif isinstance(mat_p, Mchr):\n",
    "        pm = mat_p.PM(\n",
    "            None,\n",
    "            [region, region],\n",
    "            proximity = 3\n",
    "        )\n",
    "        heatmap_title = \"Proximity Matrix\"\n",
    "    elif isinstance(mat_p, pd.DataFrame):\n",
    "        # assume all_rg df\n",
    "        samples = mat_p.columns\n",
    "        pm = ((mat_p[samples]**2).mean(axis=1)**0.5).unstack()\n",
    "        heatmap_title = \"mean RG Matrix\"\n",
    "    else:\n",
    "        raise ValueError(\"mat_p must be a path or Mchr object\")\n",
    "    # --- plot --- #\n",
    "    fig = make_subplots(\n",
    "        rows=2, cols=1,\n",
    "        shared_xaxes=True,\n",
    "        vertical_spacing=0.1,\n",
    "        row_heights=[0.8, 0.2],\n",
    "        subplot_titles=(heatmap_title, line_title)\n",
    "    )\n",
    "    heatmap = _plot_mat(\n",
    "        pm,\n",
    "        title = \"\",\n",
    "        donorm = False,\n",
    "        ignore_diags = False,\n",
    "        cmap = \"Viridis\",\n",
    "        **pm_args\n",
    "        ).data[0]\n",
    "    fig.add_trace(heatmap, row=1, col=1)\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x = prob[\"start\"],\n",
    "            y = prob[\"TAD_prob\"],\n",
    "            mode = \"lines\",\n",
    "            name = \"TAD boundary probability\",\n",
    "            line = dict(\n",
    "                color = \"black\",\n",
    "                width = 2\n",
    "            )\n",
    "        ),\n",
    "        row=2, col=1\n",
    "    )\n",
    "    fig.update_layout(\n",
    "        height= 650,\n",
    "        width = 500,\n",
    "        title = title,\n",
    "        plot_bgcolor = \"rgba(0,0,0,0)\",\n",
    "    )\n",
    "    return fig"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
